### PR TITLE
Event tracking on top stories component

### LINKED ISSUE #9082: Use view and click tracking hooks on top stories component
**Is your feature request related to a problem? Please describe.**
As a ws product manager I want to know how many users who see the top stories promos on our pages click on the links so that I have a baseline to measure improvements to the promo .

**Describe the solution you'd like**
Enable tracking to the top stories block that will send a single view event when the block is in view and a single click event if any link within the block is clicked. The content of the view and click event must be the same for the block so ATI can calculate the click/view value.  The tracking must contain details of what service is being tracked.

Configure the Top Stories block on cps articles to enable block level tracking. This will give a view in ATI where we can track the views/clicks of the block.

Attach the view event to the first item in the top stories component and a click event on each link within.

e.g. https://www.bbc.com/pidgin/world-56829018

image-2021-04-26-12-41-45-825

### The click event
Send a block level click event if any link/promo within the block is clicked. (we are only tracking the interaction and not tracking the link location)

Don't send a click event if any other part of the block is clicked.

`atc=PUB-[Campaign]-[Creation]-[Variant]-[Format]-[General Placement]-[Detailed Placement]-[Advertiser ID]-[Url]&type=AT`

```
(atc)
atc=PUB-[article-sty]-[top-stories]-[]-[]-[pidgin.world.story.56829018.page]-[]-[]-[]&type=AT
```

### The view/impression event

`ati=PUB-[Campaign]-[Creation]-[Variant]-[Format]-[General]-[Detailed]-[Advertiser]-[Url]&type=AT`

```
(ati)
ati=PUB-[article-sty]-[top-stories]-[]-[]-[pidgin.world.story.56829018.page]-[]-[]-[]&type=AT
```

### Example values from click/impression events

|                    |                                                                  |
| ------------------ | ---------------------------------------------------------------- |
| Campaign           | 'article-sty' (this is dynamic and relates to the page type)                                                        |
| Creation           | 'top-stories'                                                    |
| Variant            | empty                                                            |
| Format             | empty                                                            |
| General Placement  | _page_ eg pidgin.world.story.56829018.page (same as the p value) |
| Detailed Placement | empty                                                            |
| Advertiser ID      | empty                                                            |
| Url                | empty                                                            |

### ATI Tag Inspector
In the Tag Inspector these values should be present:
- Site - 598342
- Level2 - 70
- (Under -Ad)
- Campaign ID - article-sty
- Creation ID - top-stories
- General placement - pidgin.world.story.56829018.page

### Accessibility
The solution should have no adverse affects with any of our support assistive technology. This will require a developer accessiblity swarm to check that nothing unexpected is read out, such as 'clickable'. e.g. Only the content and HTML semantics should be announced for elements, nothing extra to the defined screen reader UX for components. Note as we are using the same technical solution, this testing should be carried out on either this Issue or https://github.com/bbc/simorgh/issues/9081 - which ever is completed first.

**Describe alternatives you've considered**

**Testing notes**
[Tester to complete]

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [x] This feature is expected to need manual testing.

**Additional context**
The WS event tracking matrix contains all tracking campaigns and which components are appear on which page types NEWSWORLDSERVICE/WS+event+matrix.

### PR DESCRIPTION
**Pull request by jroebu14**
Resolves https://github.com/bbc/simorgh/issues/9082

**Overall change:**
Adds view and click event tracking mechanisms to the top stories component.

**Technical implementation:**
Creates a new React context called `EventTrackingContext` that is composed with other Simorgh contexts in the `withContexts` high order component which wraps all page components. `EventTrackingContext` provides all child components with most of the necessary props needed to be able to send an event to ATI. Custom hooks `useViewTracker`and `useClickTrackerHandler` receive the props from the context along with specific campaign and component type based props that needs passed into each hook call. When the view and click event mechanisms are triggered then all props are passed into `sendEventBeacon` which makes the call to ATI to log the event with all of the event data.

**Code changes:**

- updates `useViewTracker` hook API and [docs](https://github.com/bbc/simorgh/blob/click-view-hook/src/app/hooks/useViewTracker/README.md)
- updates `useClickTrackerHandler` hook API and [docs](https://github.com/bbc/simorgh/blob/click-view-hook/src/app/hooks/useClickTrackerHandler/README.md)

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [x] I have assigned this PR to the Simorgh project
- [x] (BBC contributors only) This PR follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Testing:**
- Go to any story page for any service and view the top stories component and ensure a request to ATI is sent with the correct tracking values e.g.:
Screenshot 2021-05-21 at 09 14 30

- Go to any story page for any service and click on a top stories link and ensure a request to ATI is sent with the correct tracking values.
- Ensure the view event tracking data and the click event tracking data is the same with the exception of the `ati` for the view event and `atc` for the click event.
- Check Safari bug - from this issue https://github.com/bbc/simorgh/issues/9070
> 🐛 There is an issue with ATI for Safari where navigating back after clicking on a tracked link triggers the link again - story BBCHOME-398. We must consider a workaround for this. WebCore have fixed this in their implementation - refer to their workaround for this.

- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [x] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
- [x] This PR requires manual testing

### CODE REVIEW

File: src/app/containers/CpsTopStories/index.jsx [hunk 1]
```diff
@@ -5,49 +5,75 @@ import { pathOr } from 'ramda';
...
+          key={item.id || item.uri}
+          ref={isFirstItem(index) ? viewRef : null}
+        >
+          <StoryPromo
+            item={item}
+            dir={dir}
+            displayImage={false}
+            displaySummary={false}
+            serviceDatetimeLocale={serviceDatetimeLocale}
+            trackingData={EVENT_TRACKING_DATA}
```
Comment (jroebu14) [inline src/app/containers/CpsTopStories/index.jsx]: should maybe be called `eventTrackingData` instead:

```jsx
eventTrackingData={EVENT_TRACKING_DATA}
```

File: src/app/containers/StoryPromo/index.jsx [hunk 2]
```diff
@@ -108,6 +110,7 @@ const StoryPromoContainer = ({
     timezone,
   } = useContext(ServiceContext);
   const { pageType } = useContext(RequestContext);
+  const handleClickTracking = useClickTracker(trackingData);
```
Comment (jroebu14) [inline src/app/containers/StoryPromo/index.jsx]: should rename this hook to something like `useClickTrackerHandler`

File: src/app/hooks/useClickTracker/index.jsx [hunk 3]
```diff
@@ -1,96 +1,81 @@
...
     event => {
-      event.stopPropagation();
-      event.preventDefault();
-
-      if (isValidClick(event)) {
-        const nextPageUrl = href || event.target.href;
-        clickRef.current?.removeEventListener('click', handleClick);
+      if (!clicked) {
+        event.stopPropagation();
+        event.preventDefault();
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.jsx]: if we moved these lines inside of the if statement on line 35 then we wouldn't need an else which means we don't need to define the function `changeUserLocationIfDefined`

Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.jsx]: Good point, but I think moving them to inside the line 39 if statement is the one that enables us to remove the else, since the else is in case the event shouldn't be sent

File: src/app/hooks/useClickTrackerHandler/index.jsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+          sendEventBeacon({
+            type: EVENT_TYPE,
+            campaignName,
+            componentName,
+            format,
+            pageIdentifier,
+            platform,
+            service,
+            statsDestination,
+            url: window.location.href,
```
Comment (jroebu14) [inline src/app/hooks/useClickTrackerHandler/index.jsx]: we're not supposed to send this on every click event https://github.com/bbc/simorgh/issues/9082#issuecomment-839653051

File: src/app/containers/CpsOnwardJourney/index.jsx [hunk 5]
```diff
@@ -221,10 +221,14 @@ const CpsOnwardJourney = ({
         ) : null}
         {hasSingleContent ? (
           <SingleContentWrapper columnType={columnType}>
-            {promoComponent({ promo: singleContent, dir })}
```
Comment (jroebu14) [inline src/app/containers/CpsOnwardJourney/index.jsx:224]: this component being called as a function was for some reason causing infinite re-renders for us when we implemented the view tracker. Don't fully know why but we reasoned we shouldn't be doing this sort of weirdness anyway. Maybe it needs to be JSX so that react can do its reconciliation.

Comment (andrewscfc) [reply src/app/containers/CpsOnwardJourney/index.jsx:224]: hmm, I'd love to know what's going on with that 🤔

**Review by andrewscfc — COMMENTED** [review]
Just a question on the overall approach am I right in saying we track views based upon someone seeing the first top story (number 1) and then track an associated click based on any one of the stories being clicked?

Feels like it might be cleaner if we tracked the view based on the entire component being viewed, so maybe attach the view tracking to the containing TopStories component?

This PR, though appears to be vindicating seperating View and Click hooks somewhat but it just might be nice to track the view against the overal top stories component rather than one of Top Stories?

File: src/app/hooks/useClickTrackerHandler/index.jsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+          }).finally(() => {
```
Comment (andrewscfc) [inline src/app/hooks/useClickTrackerHandler/index.jsx]: could we use `try {} finally` syntax here: e.g.
```
try {
  sendEventBeacon()
} 
finally {
  if(nextPageUrl) {
    window.location.assign(nextPageUrl);
  }
}
```

Just most of `promise` style APIs can be replaced by language constructs.

File: src/app/hooks/useClickTrackerHandler/index.test.jsx [hunk 7]
```diff
@@ -187,11 +159,25 @@ describe('Click tracking', () => {
...
+      return (
+        <WithContexts pageData={pidginData}>
+          <div onClick={handleClick}>
+            <TestComponent hookProps={defaultProps} />
+          </div>
+        </WithContexts>
+      );
+    };
+
+    const { getByText } = render(<TestComponentContainer />);
```
Comment (andrewscfc) [inline src/app/hooks/useClickTrackerHandler/index.test.jsx]: lovely to see how this API is evolving 😄

File: src/app/hooks/useClickTrackerHandler/index.jsx (hunk 6 above)
Comment (jroebu14) [reply src/app/hooks/useClickTrackerHandler/index.jsx]: oh yeah good shout - it was a promise previously because it used to be inside of a useEffect

**Comment by jroebu14** [comment]
> Just a question on the overall approach am I right in saying we track views based upon someone seeing the first top story (number 1) and then track an associated click based on any one of the stories being clicked?
> Feels like it might be cleaner if we tracked the view based on the entire component being viewed, so maybe attach the view tracking to the containing TopStories component?
> This PR, though appears to be vindicating seperating View and Click hooks somewhat but it just might be nice to track the view against the overal top stories component rather than one of Top Stories?

Yep this is correct. This requirement is detailed in the issue and reasoning given in the 3 amigos. Basically @JonBeeb explained that if a mobile user scrolls to the first item and clicks it then a view event wouldn't be registered if the view hook was attached to the whole component.

File: src/app/containers/CpsOnwardJourney/index.jsx (hunk 5 above)
Comment (jroebu14) [reply src/app/containers/CpsOnwardJourney/index.jsx:224]: Too tight on time to investigate but I think we should be using JSX rather than functions to render components

**Comment by andrewscfc** [comment]
> > Just a question on the overall approach am I right in saying we track views based upon someone seeing the first top story (number 1) and then track an associated click based on any one of the stories being clicked?
> > Feels like it might be cleaner if we tracked the view based on the entire component being viewed, so maybe attach the view tracking to the containing TopStories component?
> > This PR, though appears to be vindicating seperating View and Click hooks somewhat but it just might be nice to track the view against the overal top stories component rather than one of Top Stories?
> 
> Yep this is correct. This requirement is detailed in the issue and reasoning given in the 3 amigos. Basically @JonBeeb explained that if a mobile user scrolls to the first item and clicks it then a view event wouldn't be registered if the view hook was attached to the whole component.

Hmm won't there be a bug though if you scroll down to the 10th item quickly and click it and the 1st item isn't visible? That'll give you a click without a view? I thought if any part of the Top Stories component were visible for the required amount of time you would register a view? Or does that only work if the whole component is visible?

**Comment by JonBeeb** [comment]
> Just a question on the overall approach am I right in saying we track views based upon someone seeing the first top story (number 1) and then track an associated click based on any one of the stories being clicked?
> 
> Feels like it might be cleaner if we tracked the view based on the entire component being viewed, so maybe attach the view tracking to the containing TopStories component?
> 
> This PR, though appears to be vindicating seperating View and Click hooks somewhat but it just might be nice to track the view against the overal top stories component rather than one of Top Stories?

Thought about that approach but it we have big list of promos with 7 items I'm not sure the component block will be registered (f we used the 50% rule).  I think Features has up to 9 promos so (if we use the 50% rule) it would take a lot of scrolling even on a desktop to see half of it. Happy to discuss if you think there is a better way of doing it.

File: src/app/containers/StoryPromo/index.jsx [hunk 8]
```diff
@@ -303,6 +309,12 @@ StoryPromoContainer.propTypes = {
   isRecommendation: bool,
   isSingleColumnLayout: bool,
   serviceDatetimeLocale: string,
+  eventTrackingData: shape({
+    pageIdentifier: string,
+    platform: string,
+    service: string,
```
Comment (jroebu14) [inline src/app/containers/StoryPromo/index.jsx]: need to remove this since the hooks now gets it from the service context

**Comment by jroebu14** [comment]
> Hmm won't there be a bug though if you scroll down to the 10th item quickly and click it and the 1st item isn't visible? That'll give you a click without a view? I thought if any part of the Top Stories component were visible for the required amount of time you would register a view? Or does that only work if the whole component is visible?

Good point about the fast scrolling.

It's always been the case that 50% needs to be in view to register a view. If there wasn't some sort of threshold then the user could scroll a pixel of the component into view, navigate elsewhere and a view event would be registered even though the user only had 1px of the component in view and didn't actually view it.

**Comment by andrewscfc** [comment]
> > Hmm won't there be a bug though if you scroll down to the 10th item quickly and click it and the 1st item isn't visible? That'll give you a click without a view? I thought if any part of the Top Stories component were visible for the required amount of time you would register a view? Or does that only work if the whole component is visible?
> 
> Good point about the fast scrolling.
> 
> It's always been the case that 50% needs to be in view to register a view. If there wasn't some sort of threshold then the user could scroll a pixel of the component into view, navigate elsewhere and a view event would be registered event though the user only had 1px of the component in view and didn't actually view it.

I've not thought about it very deeply but maybe the rules could be changed for 'containing elements' so if even 1pixel was visible it would be considered viewed. You pass a prop into the view tracker to inform it that the element is a containing element? Again, not very deeply thought through but the existing approach may have a bit of a flaw, maybe it's an acceptable one though?

File: src/app/hooks/useViewTracker/index.test.jsx [hunk 9]
```diff
@@ -130,19 +154,38 @@ describe('Expected use', () => {
     });
   });
 
-  it('should not send event to ATI when element is in view for less than 1 second', async () => {
+  it('should send not event to ATI when eventTracking toggle is disabled', async () => {
```
Comment (jroebu14) [inline src/app/hooks/useViewTracker/index.test.jsx]:
```suggestion
  it('should not send event to ATI when eventTracking toggle is disabled', async () => {
```

**Comment by jroebu14** [comment]
The 1px idea has a flaw too in that we could potentially send a lot of false view events.

I'll have a think too about a better solution.

**Comment by jroebu14** [comment]
@JonBeeb @andrewscfc I can easily change it so that if any of the items are 50% in view then we register only 1 view event

**Comment by JonBeeb** [comment]
> > > Hmm won't there be a bug though if you scroll down to the 10th item quickly and click it and the 1st item isn't visible? That'll give you a click without a view? I thought if any part of the Top Stories component were visible for the required amount of time you would register a view? Or does that only work if the whole component is visible?
> > 
> > 
> > Good point about the fast scrolling.
> > It's always been the case that 50% needs to be in view to register a view. If there wasn't some sort of threshold then the user could scroll a pixel of the component into view, navigate elsewhere and a view event would be registered event though the user only had 1px of the component in view and didn't actually view it.
> 
> I've not thought about it very deeply but maybe the rules could be changed for 'containing elements' so if even 1pixel was visible it would be considered viewed. You pass a prop into the view tracker to inform it that the element is a containing element? Again, not very deeply thought through but the existing approach may have a bit of a flaw, maybe it's an acceptable one though?

The 50% for 1 sec comes from google (via webcore) for tracking the impression, I assume to stop the capturing an impression when it has not really been seen, but you are right, someone could fast scroll down the page and it not be captured if we are only looking at the first item (although I can't see that as being the normal use case for someone reading an article.).

**Comment by JonBeeb** [comment]
> @JonBeeb @andrewscfc I can easily change it so that if any of the items are 50% in view then we register only 1 view event

That would work?

**Comment by jroebu14** [comment]
> > @JonBeeb @andrewscfc I can easily change it so that if any of the items are 50% in view then we register only 1 view event
> 
> That would work?

Yep, quite easily thanks to React hooks. I think @HarryVerhoef and I should organise a quick demo and mob code review when this PR is ready. We'll send you an invite.

File: src/app/contexts/EventTrackingContext/index.jsx [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error(
+        `ATI Event Tracking Error: Could not parse tracking values from page data:\n${error.message}`,
+      );
+    }
+  }
+
+  const { pageType } = requestContext;
+  const campaignName = {
```
Comment (jroebu14) [inline src/app/contexts/EventTrackingContext/index.jsx]: I might rename this to `campaignID`

File: src/app/hooks/useClickTrackerHandler/index.jsx [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+              componentName,
+              format,
+              pageIdentifier,
+              platform,
+              service,
+              statsDestination,
+            });
+          } finally {
+            if (nextPageUrl) {
+              window.location.assign(nextPageUrl);
```
Comment (jroebu14) [inline src/app/hooks/useClickTrackerHandler/index.jsx:58]: is this a security risk? should we use react-router-dom redirect component?

File: src/app/hooks/useViewTracker/index.jsx [hunk 12]
```diff
@@ -1,50 +1,38 @@
-import { useEffect, useState, useContext, useRef } from 'react';
+import { useContext, useEffect, useState, useRef } from 'react';
+import path from 'ramda/src/path';
 import { useInView } from 'react-intersection-observer';
 // Polyfill IntersectionObserver, e.g. for IE11
 import 'intersection-observer';
```
Comment (jroebu14) [inline src/app/hooks/useViewTracker/index.jsx:5]: look into removing this

File: src/app/containers/ATIAnalytics/README.md [hunk 13]
```diff
@@ -6,188 +6,17 @@ This directory contains code for logging events to ATI Analytics.
...
-      componentData,
-    });
-
-    sendEventBeacon({
-      componentName,
-      type: 'click',
-      componentInfo,
-      ...eventTrackingProps,
-    });
+For logging click events you will need four things:
```
Comment (ryanmccombe) [inline src/app/containers/ATIAnalytics/README.md]: This is probably out of date (only one bullet point, not four)

File: src/app/hooks/useClickTrackerHandler/index.jsx (hunk 11 above)
Comment (jroebu14) [reply src/app/hooks/useClickTrackerHandler/index.jsx:58]: @HarryVerhoef are you able to look into this?

Comment (HarryVerhoef) [reply src/app/hooks/useClickTrackerHandler/index.jsx:58]: @jroebu14 Yeah I can try and find something 👍

File: src/app/contexts/EventTrackingContext/index.jsx [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+  let pageIdentifier;
+  let platform;
+  let statsDestination;
+
+  const requestContext = useContext(RequestContext);
+  const serviceContext = useContext(ServiceContext);
+  const { enabled: eventTrackingIsEnabled } = useToggle('eventTracking');
+
+  if (pageData && eventTrackingIsEnabled) {
+    try {
```
Comment (ryanmccombe) [inline src/app/contexts/EventTrackingContext/index.jsx]: Is this try/catch just for logging?  This function seems to continue even if `buildATIClickParams` throws?

Maybe the exception handling / logging makes more sense inside `buildATIClickParams` rather than here?  Would simplify this function a bit

File: src/app/contexts/EventTrackingContext/index.jsx [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    [MOST_WATCHED_PAGE]: 'list-datadriven-watched',
+    [INDEX_PAGE]: 'index-section',
+    [FEATURE_INDEX_PAGE]: 'index-section-fix',
+    [MEDIA_ASSET_PAGE]: 'article-media-asset',
+    [STORY_PAGE]: 'article-sty',
+    [PHOTO_GALLERY_PAGE]: 'article-photo-gallery',
+    [CORRESPONDENT_STORY_PAGE]: 'article-csp',
+  }[pageType];
+
+  if (!campaignName && pageData) {
```
Comment (ryanmccombe) [inline src/app/contexts/EventTrackingContext/index.jsx]: I think this function might be a bit easier to follow if it handled some of these error cases with an early return - eg, if `pageData` or some other critical stuff is not defined, we could just return right at the start of the function with a `value` of `{}` as that will be the inevitable result anyway

File: src/app/hooks/useViewTracker/README.md [hunk 16]
```diff
@@ -5,39 +5,152 @@ The `useViewTracker` hook handles:
...
-The hook returns a ref that can be assigned to the DOM element you want to monitor for a view/impression event that will be sent to ATI.
+`useViewTracker` returns a ref that can be assigned to the DOM element you want to monitor for a view event. When a view event is triggered then the hook will send the event data to ATI.
+
+`useViewTracker` will only send 1 view event per hook initialisation. In other words, when visiting a page containing a component that is tracking views and a view event is logged, then it is only logged once (per page visit) no matter how many times the user scrolls the component in and out of view.
+
+### Props
+
+| Argument      | Type   | Required | Example                                                                                                          |
+| ------------- | ------ | -------- | ---------------------------------------------------------------------------------------------------------------- |
+| componentName | string | yes      | The name of the component or an url encoded title of a promo e.g. `most_read` or `This%20is%20a%20promo%20title` |
```
Comment (ryanmccombe) [inline src/app/hooks/useViewTracker/README.md]:
```suggestion
| componentName | string | yes      | The name of the component or a url encoded title of a promo e.g. `most_read` or `This%20is%20a%20promo%20title` |
```

File: src/app/hooks/useViewTracker/README.md [hunk 17]
```diff
@@ -5,39 +5,152 @@ The `useViewTracker` hook handles:
...
+| url           | string | no       | The url of the page e.g. `https://www.bbc.com/mundo/noticias-america-latina-56989232`                            |
 
 ### Usage
 
+**It's important to note that `useViewTracker` logs only 1 view event per instance.**
+
+With this in mind, here are some examples of how you could setup view tracking for a component:
+
+1. Log 1 view event when a component is viewed. Refer to the `Promo` component in the example below.
+2. Log 1 view event when any of the the component's containing items are viewed. Refer to the `TopStories` component in the example below.
```
Comment (ryanmccombe) [inline src/app/hooks/useViewTracker/README.md]:
```suggestion
2. Log 1 view event when any of the component's containing items are viewed. Refer to the `TopStories` component in the example below.
```

**Review by ryanmccombe — APPROVED** [review]
👍

File: src/app/contexts/EventTrackingContext/index.jsx (hunk 15 above)
Comment (jroebu14) [reply src/app/contexts/EventTrackingContext/index.jsx]: Yeh but unfortunately that's not possible because then the useContexts hooks would be called conditionally and that is not recommended by React

File: src/app/contexts/EventTrackingContext/index.jsx (hunk 14 above)
Comment (jroebu14) [reply src/app/contexts/EventTrackingContext/index.jsx]: Continuing on and not breaking the app is the behaviour we want in the event the pageData is not as expected.

Not sure if moving the exception handling to another function improves anything. Could do though

**Comment by greenc05** [comment]
**A11y swarm**

- [x] Axe
- [x] bbc-a11y

_AT Mobile:_ 
- [x] TalkBack (Chrome on Android)
- [x] VoiceOver iOS (iPhone/iPad)

_AT Desktop:_ 
- [x] Read&Write 
- [x] ZoomText Magnifier/Reader
- [x] JAWS
- [x] NVDA
- [x] VoiceOver OS (Mac)

File: src/app/containers/ATIAnalytics/README.md [hunk 18]
```diff
@@ -4,190 +4,19 @@ This directory contains code for logging events to ATI Analytics.
 
 ## Logging Click Events
 
-For logging click events you will need four things:
+There is a custom within Simorgh that handles logging click events.
```
Comment (andrewscfc) [inline src/app/containers/ATIAnalytics/README.md]:
```suggestion
There is a custom hook within Simorgh that handles logging click events.
```

File: src/app/containers/ATIAnalytics/README.md [hunk 19]
```diff
@@ -4,190 +4,19 @@ This directory contains code for logging events to ATI Analytics.
...
-## How to set up tracking for a component
+You can find instructions and examples on how to set up and use this custom hook [here](https://github.com/bbc/simorgh/blob/latest/src/app/hooks/useClickTrackerHandler/README.md).
 
-In the container for the component you're trying to track, import `EventContext`, and use `useClickTracker` as such:
+## Logging View Events
 
-```jsx
-import React, { useContext } from 'react';
-import { EventContext } from '#contexts/EventContext';
+There is a custom within Simorgh that handles logging view events.
```
Comment (andrewscfc) [inline src/app/containers/ATIAnalytics/README.md]:
```suggestion
There is a custom hook within Simorgh that handles logging view events.
```

File: src/app/containers/ATIAnalytics/README.md [hunk 20]
```diff
@@ -4,190 +4,19 @@ This directory contains code for logging events to ATI Analytics.
...
 
-const componentContainer = () => {
-  const { useClickTracker } = useContext(EventContext);
+- [useClickTrackerHandler](https://github.com/bbc/simorgh/blob/latest/src/app/hooks/useViewTracker/index.jsx) is a custom hook used that returns a `ref` that you can assign to DOM elements you'd like to view track.
```
Comment (andrewscfc) [inline src/app/containers/ATIAnalytics/README.md]: useClickTrackerHandler? Seems to link to view tracking?

File: src/app/contexts/EventTrackingContext/index.test.jsx [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+});
+
+const TestComponent = () => {
+  const trackingData = useContext(EventTrackingContext);
+
+  return <div data-testid="test-component">{JSON.stringify(trackingData)}</div>;
+};
+
+describe('Expected use', () => {
+  it('should provide tracking data to all child component', () => {
```
Comment (andrewscfc) [inline src/app/contexts/EventTrackingContext/index.test.jsx]:
```suggestion
  it('should provide tracking data to all child components', () => {
```

File: src/app/contexts/EventTrackingContext/index.test.jsx [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    const testEl = screen.getByTestId('test-component');
+    const trackingData = JSON.parse(testEl.textContent);
+
+    expect(trackingData).toEqual({});
+    expect(errorMessage).toBeUndefined();
+    expect(console.error).toHaveBeenCalledWith(
+      expect.stringContaining(
+        "ATI Event Tracking Error: Could not get the page type's campaign name",
```
Comment (andrewscfc) [inline src/app/contexts/EventTrackingContext/index.test.jsx:166]: Further to your comment above, can see campaign name and campaign id being used interchangibly, may be good to pick one for readability?

File: src/app/hooks/useClickTrackerHandler/README.md [hunk 23]
```diff
@@ -0,0 +1,130 @@
+# useClickTracker hook
```
Comment (andrewscfc) [inline src/app/hooks/useClickTrackerHandler/README.md:1]: Should we add a note here to say that the click tracker should always be used in the context of tracked view? I agree with the components themselves not being combined together but it is important that engineers know that a click isn't very useful to Product without a view I think?

**Review by andrewscfc — APPROVED** [review]
Looks great! Few docs comments and that `location.assign` issue would be good to resolve, though without user input I can't immediately think of the risk it poses?

**Comment by HarryVerhoef** [comment]
Safari webcore issue:
If a user clicks a tracked component and is taken to another page, if they go 'back' then the tracking mechanism sends another request to ATI.

I don't see this behaviour on this branch on Safari desktop, but instead after going 'back' I can view and click the same link but an ATI request is NOT sent for either view or click events (one is sent for page view). This behaviour doesn't happen on chrome so I can only assume it's some Safari weirdness. Currently setting up an iPhone emulator to see if this and/or the webcore issue occurs on iOS Safari.

~~[EDIT]: This could potentially be to do with the way we redirect users. Perhaps `window.location.assign` is the cause of this inconsistency.~~

File: src/app/hooks/useClickTrackerHandler/index.jsx (hunk 11 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTrackerHandler/index.jsx:58]: @andrewscfc @jroebu14 I think using react-router we'd only be able to redirect to a relative url, and so wouldn't be able to do an external redirect.

File: src/app/contexts/EventTrackingContext/index.jsx (hunk 14 above)
Comment (ryanmccombe) [reply src/app/contexts/EventTrackingContext/index.jsx]: I mean moving it into the `buildATIClickParams` function - the only thing in this `try` block here is a call to that function.  If that function is expected to throw exceptions, maybe that function should catch them, rather than dealing with it in this (already very complex) function?

Either way is fine though - just my first reaction to this function is that it's quite hard to follow - 70 lines and a lot of conditionals

File: src/app/contexts/EventTrackingContext/index.jsx (hunk 10 above)
Comment (ryanmccombe) [reply src/app/contexts/EventTrackingContext/index.jsx]: Could also be a separate function, to reduce the amount of things happening here

`const campaignId = getCampaignId(requestContext)`

Which could also absorb the conditional logic and logging from line 64

File: src/app/hooks/useClickTrackerHandler/index.jsx (hunk 11 above)
Comment (andrewscfc) [reply src/app/hooks/useClickTrackerHandler/index.jsx:58]: ~Do we have any need to support external links? I'd say would be good to use react router if not, if we do I guess stick with existing approach?~
As discussed in standup, no need for a change here

File: src/app/contexts/EventTrackingContext/index.jsx (hunk 14 above)
Comment (jroebu14) [reply src/app/contexts/EventTrackingContext/index.jsx]: @ryanmccombe thanks for your thoughts. I've moved the exception handling to the param building function and tried to simplify the event tracking context. Let me know if this is better

File: src/app/hooks/useClickTrackerHandler/README.md (hunk 23 above)
Comment (jroebu14) [reply src/app/hooks/useClickTrackerHandler/README.md:1]: @andrewscfc does this look OK? https://github.com/bbc/simorgh/pull/9101/commits/1db6ac77035d4d98a16e8c8881d355bca33062ff#diff-fb7721b3aeed707492606afb9ed4233e97cbbe722ffb629f0c0bb7428522e7ac
